README.md

# Mango - Fastlane plugin

[![fastlane Plugin Badge](https://rawcdn.githack.com/fastlane/fastlane/master/fastlane/assets/plugin-badge.svg)](https://rubygems.org/gems/fastlane-plugin-mango) [![License](https://img.shields.io/badge/license-MIT-green.svg?style=flat)](https://github.com/xing/mango/blob/master/LICENSE)
[![Gem](https://img.shields.io/gem/v/fastlane-plugin-mango.svg?style=flat)](http://rubygems.org/gems/fastlane-plugin-mango)

A fastlane plugin that runs Android tasks on a specified [Docker](https://www.docker.com/) image

<img src="assets/mango_logo.png" alt="Mango Logo" width="256px" height="256px"/>

Running Android tests, especially [Espresso](https://developer.android.com/training/testing/espresso/) on a continuous integration environment like [Jenkins](https://jenkins.io/) can be a hassle. You need to boot, manage and destroy an [Android Virtual Device (AVD)](https://developer.android.com/studio/run/managing-avds) during the test run. This is why we, the mobile releases team at [XING](https://www.xing.com), built this plugin. It spins up a specified [Docker](https://www.docker.com/) image and runs a given task on it.

Another requirement we had was to run on a clean environment, which is why Mango also helps us to run our unit test and more. For an example check out the [example `Fastfile`](sample-android/fastlane/Fastfile)

## Prerequisites

In order to use this plugin you will need to to have [Docker](https://www.docker.com/) installed on the machine you are using.
Documentation on how to set it up properly on a Linux (ubuntu) machine can be found [here](docs/docker-linux.md).

If you need an Android Virtual Device (AVD) to run your tests (for example Espresso, Calabash or Appium), it's necessary to check that your CPU supports kvm virtualisation. We already experienced that it doesn't fully work on macOS and are using Linux for that.

## Getting Started

This project is a [_fastlane_](https://github.com/fastlane/fastlane) plugin. To get started with `fastlane-plugin-mango`, add it to your project by running:

```bash
fastlane add_plugin mango
```

## Usage

After installing this plugin you have access to one additional action (`mango`) in your `Fastfile`.

So a lane in your `Fastfile` could look similar to this for Espresso tests:
```ruby
desc "Run espresso tests on docker images"
  lane :Espresso_Tests do
     run_dockerized_task(
       container_name: "espresso_container",
       port_factor: options[:port_factor],
       docker_image: "joesss/mango-docker:latest",
       container_timeout: 120,
       android_task: "./gradlew connectedAndroidTest",
       post_actions: "adb logcat -d > logcat.txt",
       pull_latest_image: true          
     )
   end
```

or to this for unit tests or other gradle tasks:
```ruby
desc "Run unit tests on docker images"
  lane :Unit_Tests do
     run_dockerized_task(
       container_name: "unit_tests_container",
       port_factor: options[:port_factor],
       docker_image: "joesss/mango-base:latest",
       is_running_on_emulator: false,
       android_task: "./gradlew testDebug",
       pull_latest_image: true         
     )
   end
```

Now you can call this new lane by calling `bundle exec fastlane Espresso_Tests`.

The Plugin will start up the given `docker_image`, execute the given `android_task` and afterwards execute the `post_actions`.

## Configuration options
The `mango` action has plenty of options to configure it.

| Option | Description | Default value | Optional | Type |
| - |:-|-:| :-:| -:|
| `container_name`| Name of the docker container. Will be generated randomly if not defined. | - | ✅ | `String` |
| `no_vnc_port` | Port to redirect noVNC. To observe your docker container from your browser while it's running. | 6080 | ❌ | `Integer` |
| `device_name` | Name of the Android device. | Nexus 5X | ❌ | `String` |
| `emulator_name`| Name of the Android emulator. | emulator-5554 | ❌ | `String` |
| `docker_image` | Name of the Docker image, that should be started and used to run your tasks. | butomo1989/docker-android-x86-5.1.1 | ❌ | `String` |
| `container_timeout` | Timeout (in seconds) to get a healthy docker container. Depending on your `docker_image` it may take some time until it's started up and ready to use. | 450 (this equals 7.5 minutes) | ❌ | `Integer` |
| `android_task` | A generic Android task you want to execute. | - | ❌ | `String` |
| `sdk_path` | The path to your Android sdk directory. | `ANDROID_HOME` environment variable | ✅ | `String` |
| `port_factor` | Base for calculating a unique port for noVNC. We recommend to use the `EXECUTOR_NUMBER` from your Jenkins environment. | - | ✅ | `String` |
| `workspace_dir` | Path to the workspace to execute commands. If you want to execute your `android_task` from a different directory you have to specify `workspace_dir`. | `/root/tests/` | ✅ | `String` |
| `maximal_run_time` | Defines the maximal time of your test run in minutes. This can be helpful if you want to kill hanging processes automatically after a certain time. | 60 | ✅ | `Integer` |
| `bundle_install` | Defines if the Android task must execute `bundle install` before running a build. This is useful if you want to execute non-gradle tasks. Like [Calabash](https://github.com/calabash/calabash-android), where you need to update/install your [Ruby](https://www.ruby-lang.org) dependencies. | `true` | ✅ | `Boolean` |
| `is_running_on_emulator` | Define if container boots up an emulator instance inside of it. This will trigger configuration of noVNC, assign necessary ports etc. | `true` | ✅ | `Boolean` |
| `post_actions` | Actions that will be executed after the main command has been executed. | - | ✅ | `String` |
| `pre_action` | Actions that will be executed before the docker image gets pulled | - | ✅ | `String` |
| `docker_registry_login` | Command to authenticate yourself in a custom Docker registry | - | ✅ | `String` |
| `pull_latest_image` | Defines if you want to pull the `:latest` tag of the given `docker_image` | `false` | ✅ | `Boolean` |

## Contributing

🎁 Bug reports and pull requests for new features are most welcome!

👷🏼 We are looking forward to your pull request, we'd love to help!

You can help by doing any of the following:

- Reviewing pull requests
- Bringing ideas for new features
- Answering questions on issues
- Improving documentation

This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [Contributor Covenant code](http://contributor-covenant.org/) of conduct.

## Issues and Feedback

For any other issues and feedback about this plugin, please submit it to this repository.

## Troubleshooting

If you have trouble using plugins, check out the [Plugins Troubleshooting](https://docs.fastlane.tools/plugins/plugins-troubleshooting/) guide.

## License

The fastlane plugin is available as open source under the terms of the [MIT License](http://opensource.org/licenses/MIT).

## Example

Check out the [example `Fastfile`](sample-android/fastlane/Fastfile) to see how to use this plugin.
Try it by cloning the repo, running `fastlane install_plugins` and `bundle exec fastlane test`.

## Run tests for this plugin

To run both the tests and code style validation, run

```
rake
```

To automatically fix many of the styling issues, use
```
rubocop -a
```

## Using _fastlane_ Plugins

For more information about how the `fastlane` plugin system works, check out the [Plugins documentation](https://docs.fastlane.tools/plugins/create-plugin/).

## About _fastlane_

_fastlane_ is the easiest way to automate beta deployments and releases for your iOS and Android apps. To learn more, check out [fastlane.tools](https://fastlane.tools).


Changelog.md
# 1.1.4

## Fixed

- Crash when adb is not ready to work yet

# 1.1.3

## Fixed

- Crash when class instances can be unitialized


lib/fastlane/plugin/emulator_docker/actions/emulator_docker_action.rb
module Fastlane
  module Actions
    class EmulatorDockerAction < Action
      def self.run(params)
        UI.important("The emulator_docker plugin is working!")
        emulator_name = params[:emulator_name]
        docker_emulator = Fastlane::Helper::EmulatorDockerHelper.new(params)
        docker_emulator.setup_container

        failure_buffer_timeout = 5
        timeout_command = "timeout #{params[:maximal_run_time] - failure_buffer_timeout}m"
        workspace_dir = params[:workspace_dir]

        android_task = params[:android_task]
        if android_task
          UI.success("Starting Android Task.")
          bundle_install = params[:bundle_install] ? '&& bundle install ' : ''

          docker_emulator.docker_exec("cd #{workspace_dir} #{bundle_install}&& #{timeout_command} #{android_task} || exit 1")
        end

      ensure
        post_actions = params[:post_actions]
        if post_actions
          docker_emulator&.docker_exec("cd #{workspace_dir} && #{post_actions}")
        end

        UI.important("Cleaning up #{emulator_name} container")
        docker_emulator.clean_container if docker_emulator.instance_variable_get('@container')
      end

      def self.description
        "This plugin handles Android multiple emulator using docker images"
      end

      def self.authors
        ["Serghei Moret"]
      end

      def self.return_value
        # If your method provides a return value, you can describe here what it does
      end

      def self.is_supported?(platform)
        platform == :android
      end

      def self.details
        # Optional:
        "TBD"
      end

      def self.available_options
        [
          FastlaneCore::ConfigItem.new(key: :container_name,
                                      env_name: "CONTAINER_NAME",
                                      description: "Name of the docker container. Will be generated randomly if not defined",
                                      optional: true,
                                      type: String),

          FastlaneCore::ConfigItem.new(key: :no_vnc_port,
                                       env_name: "NO_VNC_PORT",
                                       description: "Port to redirect noVNC. 6080 by default",
                                       default_value: 6080,
                                       optional: false,
                                       type: Integer),

          FastlaneCore::ConfigItem.new(key: :device_name,
                                       env_name: "DEVICE_NAME",
                                       description: "Name of the Android device. Nexus 5X by default",
                                       default_value: 'Nexus 5X',
                                       optional: false,
                                       type: String),

          FastlaneCore::ConfigItem.new(key: :emulator_name,
                                       env_name: "EMULATOR_NAME",
                                       description: "Name of the Android emulator. emulator-5554 by default",
                                       default_value: 'emulator-5554',
                                       optional: false,
                                       type: String),

          FastlaneCore::ConfigItem.new(key: :docker_image,
                                       env_name: "DOCKER_IMAGE",
                                       description: "Name of the Docker image. butomo1989/docker-android-x86-5.1.1 by default",
                                       default_value: 'butomo1989/docker-android-x86-5.1.1',
                                       optional: false,
                                       type: String),

          FastlaneCore::ConfigItem.new(key: :container_timeout,
                                      env_name: "CONTAINER_TIMEOUT",
                                      description: "Timeout (in seconds) to get the healthy docker container. 450 (7.5 minutes) by default",
                                      default_value: 450,
                                      optional: false,
                                      type: Integer),

         FastlaneCore::ConfigItem.new(key: :android_task,
                                      env_name: "ANDROID TASK",
                                      description: "A generic Android task you want to execute",
                                      is_string: true,
                                      optional: false),

          FastlaneCore::ConfigItem.new(key: :sdk_path,
                                       env_name: "SDK_PATH",
                                       description: "The path to your Android sdk directory (root). ANDROID_HOME by default",
                                       default_value: ENV['ANDROID_HOME'],
                                       is_string: true,
                                       optional: true),

          FastlaneCore::ConfigItem.new(key: :port_factor,
                                      env_name: "PORT_FACTOR",
                                      description: "Base for calculating a unique port for noVNC. You can pass EXECUTOR_NUMBER from Jenkins for example, this will be unique and not clash in case of several instances running on the same machine",
                                      optional: true,
                                      type: String),

          FastlaneCore::ConfigItem.new(key: :workspace_dir,
                                       env_name: "WORKSPACE_DIR",
                                       default_value: '/root/tests/',
                                       description: "Path to the workspace to execute commands",
                                       optional: true,
                                       type: String),

          FastlaneCore::ConfigItem.new(key: :maximal_run_time,
                                       env_name: "MAXIMAL_RUN_TIME",
                                       default_value: 60,
                                       description: "Defines the maximal time of your test run. Defaults to 60 minutes",
                                       optional: true,
                                       type: Integer),

          FastlaneCore::ConfigItem.new(key: :bundle_install,
                                       env_name: "BUNDLE_INSTALL",
                                       default_value: false,
                                       description: "Defines if the Android task must execute bundle install before running a build",
                                       optional: true,
                                       type: Boolean),

          FastlaneCore::ConfigItem.new(key: :is_running_on_emulator,
                                       env_name: "IS_RUNNING_ON_EMULATOR",
                                       default_value: true,
                                       description: "Define if we want to run the emulator in the container",
                                       optional: true,
                                       type: Boolean),

          FastlaneCore::ConfigItem.new(key: :post_actions,
                                      env_name: "POST_ACTIONS",
                                      description: "Actions that will be executed after the main command has been executed",
                                      is_string: true,
                                      optional: true),

          FastlaneCore::ConfigItem.new(key: :pre_action,
                                      env_name: "PRE_ACTION",
                                      description: "Actions that will be executed before the docker image gets pulled",
                                      is_string: true,
                                      optional: true),

          FastlaneCore::ConfigItem.new(key: :docker_registry_login,
                                      env_name: "DOCKER_REGISTRY_LOGIN",
                                      description: "Authenticating yourself to a custom Docker registry",
                                      type: String,
                                      optional: true),

          FastlaneCore::ConfigItem.new(key: :pull_latest_image,
                                      env_name: "PULL_LATEST_IMAGE",
                                      description: "Define if you want to pull the latest image",
                                      type: Boolean,
                                      default_value: false,
                                      optional: true)
        ]
      end
    end
  end
end
